src/graph.rs

```diff
@@ -284,56 +285,24 @@ impl PyGraph {
                 .downcast::<PyTuple>()
                 .unwrap();
 
-            // use a pointer to iter the node list
-            let mut pointer = 0;
-            let mut next_node_idx: usize = nodes_lst
-                .get_item(pointer)
-                .unwrap()
-                .downcast::<PyTuple>()
-                .unwrap()
-                .get_item(0)
-                .unwrap()
-                .downcast::<PyLong>()
-                .unwrap()
-                .extract()
-                .unwrap();
-
             // list of temporary nodes that will be removed later to re-create holes
             let node_bound_1: usize = last_item.get_item(0).unwrap().extract().unwrap();
             let mut tmp_nodes: Vec<NodeIndex> =
                 Vec::with_capacity(node_bound_1 + 1 - nodes_lst.len());
 
-            for i in 0..nodes_lst.len() + 1 {
-                if i < next_node_idx {
+            for item in nodes_lst {
+                let item = item.downcast::<PyTuple>().unwrap();
+                let next_index: usize = item.get_item(0).unwrap().extract().unwrap();
+                let weight: PyObject = item.get_item(1).unwrap().extract().unwrap();
+                while next_index > self.graph.node_bound() {
```

Comment:
What's the purpose of this change? (This review may be mostly questions and same comments will apply to graph.rs and digraph.rs.)